core/internal/api/api.go

```diff
@@ -54,22 +74,96 @@ type Request struct {
 	Headers map[string]string
 }
 
+type BackendOptions struct {
+	// The scheme and hostname for contacting the server, not including a final
+	// slash. For example, "http://localhost:8080".
+	BaseURL *url.URL
+
+	// Logger for HTTP operations.
+	Logger *slog.Logger
+
+	// W&B API key.
+	APIKey string
+}
+
 // Creates a [Backend].
 //
 // The `baseURL` is the scheme and hostname for contacting the server, not
 // including a final slash. Example "http://localhost:8080".
-func New(baseURL *url.URL) *Backend {
-	return &Backend{baseURL}
+func New(opts BackendOptions) *Backend {
+	return &Backend{
+		opts.BaseURL,
+		opts.Logger,
+		opts.APIKey,
+	}
+}
+
+type ClientOptions struct {
+	// Maximum number of retries to make for retryable requests.
+	RetryMax int
+
+	// Minimum time to wait between retries.
+	RetryWaitMin time.Duration
+
+	// Maximum time to wait between retries.
+	RetryWaitMax time.Duration
+
+	// Function that determines whether to retry based on the response.
+	//
+	// If nil, then retries are made on connection errors and server errors
+	// (HTTP status code >=500).
+	RetryPolicy retryablehttp.CheckRetry
+
+	// Timeout for HTTP requests.
+	//
+	// This is the time to wait for an individual HTTP request to complete
+	// before considering it as failed. It does not include retries: each retry
+	// starts a new timeout.
+	NonRetryTimeout time.Duration
+
+	// Additional headers to pass in each request.
+	ExtraHeaders map[string]string
 }
 
 // Creates a new [Client] for making requests to the [Backend].
-//
-// TODO: For now this accepts a [retryablehttp.Client] directly, but it will
-// be rewritten to accept a list of specific options instead. Do not rely on
-// having a reference to the underlying client.
-func (backend *Backend) NewClient(retryableHTTP *retryablehttp.Client) *Client {
-	return &Client{
-		backend:       backend,
-		retryableHTTP: retryableHTTP,
+func (backend *Backend) NewClient(opts ClientOptions) Client {
+	retryableHTTP := clients.NewRetryClient(
+		clients.WithRetryClientBackoff(clients.ExponentialBackoffWithJitter),
+		clients.WithRetryClientRetryMax(opts.RetryMax),
+		clients.WithRetryClientRetryWaitMin(opts.RetryWaitMin),
+		clients.WithRetryClientRetryWaitMax(opts.RetryWaitMax),
+		clients.WithRetryClientHttpTimeout(opts.NonRetryTimeout),
+	)
+
+	if opts.RetryPolicy != nil {
+		retryableHTTP.CheckRetry = opts.RetryPolicy
+	}
+
+	if backend.logger != nil {
+		retryableHTTP.ResponseLogHook = hookLogErrors(backend.logger)
+		retryableHTTP.Logger = slog.NewLogLogger(
+			backend.logger.Handler(),
+			slog.LevelDebug,
+		)
+	}
+
+	return &clientImpl{backend, retryableHTTP, opts.ExtraHeaders}
+}
+
+// Returns a hook that logs all HTTP error responses.
+func hookLogErrors(logger *slog.Logger) retryablehttp.ResponseLogHook {
+	return func(_ retryablehttp.Logger, response *http.Response) {
+		if response.StatusCode < 400 {
+			return
+		}
+
+		body, err := io.ReadAll(response.Body)
+		if err != nil {
+			logger.Info(
```

Comment:
should this one be an Error status?

Reply:
good point! I was copying what we had before, but Error makes more sense than Info to me.

something to note is that this will print the error for each retry, whereas maybe we only care about the final error. What do you think?

Reply:
yeah, hmmm... good point. Ideally we have debug for each retry and for the final it is error?

Reply:
Updated this to log at the Debug level when retrying an HTTP error, and at the Error level when the final request fails.